### PR TITLE
Add a molnums topology attribute to TPR parser

### PR DESCRIPTION
**Pull request by jbarnoud**
PR #1557 introduced the moltype attribute that gives a name to each kind
of molecule in a topology. This commit adds the related molnums
attribute that gives an index to each individual molecules.

The molnums attributes allows to split a system by molecule without
ambiguity when the attribute is defines. Without this commit, the only
way of splitting a system per molecule is `atom_group.fragments`. While
it works most of the time, it can fail in some topology that use other
topology tricks than bonds to connect atoms; this is the case in gromacs
with virtual sites for instance. Fragments can also be time consuming on
large systems. This commits allows to slit a system per molecule without
ambiguity with `atom_group.groupby('molnums')` or
`atom_group.split('molecule')`.

Using `AtomGroup.split` rather than `AtomGroup.fragments` is about 10 times faster according to [a very crude benchmark](https://gist.github.com/jbarnoud/de00e574591e7b8de74e5a95c3a774a4).

This commit also adds the `molnum` selection keyword. To do so, this
commit refactored `mda.core.selection.SelectionRange` to avoid code
duplication.

Note that, like `moltypes`, `molnums` is only implemented to the TPR parser; but it could be implemented for any topology format that includes the information. Latter on, a guesser for `molnums` could be implemented based on fragments.

For some context, I need to count cholesterol molecules in my simulations. Yet, martini cholesterol defines 2 beads as virtual sites. As a result each molecule appears 3 fragments. Also, systems tend to become larger and larger, so that computing the fragments started to be an annoyance. Cholesterol is a single residue per molecule, so I fell back to counting residues, but at the same time a colleague of mine wants to count molecules in a system that contains proteins. Earlier this year, an other colleague wanted to count polymer chains and had to find a workaround. All of this was frustrating since we all use gromacs that has the molecule information.

PR Checklist
------------
 - [X] Tests?
 - [X] Docs?
 - [x] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

**Comment by richardjgowers** [comment]
Looks good.  I was thinking maybe it's confusing to have both `fragments` and `molecules`.  Would it be better to have `fragments` be guessed off bonding information (as is currently implemented) unless a topology has provided this information (as this implements).  Or is there a subtle difference in frags/mols somewhere?

**Comment by jbarnoud** [comment]
`fragments` and molecules should be the same in most cases. Currently there is at least one case when it is not the same, though: virtual sites in gromacs. Virtual sites are particles that are built based on the positions of other particles; they are "connected" to the molecule as their positions depends on the molecule, but they may not have a bond connecting them. As a result it is one molecule, but multiple fragments. This is more of an annoyance in my context, so in this case it may be relevant to use the molecule information to get the fragments if available. I can see that you may not want that if you prune bonds.

One example of the latter case would be if you work with an elastic network. Then you may want to remove some bonds based on some criterion and write a PDB with a connect record out of it. Keeping fragments the way it is now would allow to check if the resulting molecule is still connected.

All of this to say: it may be a good idea, but I am not sure about it, yet.

By the way, on a different topic: kudos on the new topology system. Implementing molnums (like implementing moltypes) was super easy!

File: package/MDAnalysis/core/groups.py [hunk 1]
```diff
@@ -1968,20 +1968,27 @@ def split(self, level):
...
+                     'molecule': 'molnums'}
 
         if level == "atom":
             return [self.universe.atoms[[a.ix]] for a in self]
 
         # higher level groupings
         try:
             levelindices = getattr(self, accessors[level])
+        except AttributeError:
+            raise ValueError('This universe does not have {} '
```
Comment (richardjgowers) [inline package/MDAnalysis/core/groups.py]: maybe this should also be an AttErr?  It's a missing attribute causing the problem

File: testsuite/MDAnalysisTests/core/test_atomselections.py [hunk 2]
```diff
@@ -415,26 +415,39 @@ def test_cyzone(self):
         assert_equal(len(sel), 1556)
 
 
-class TestSelectionsTPR(TestCase):
-    def setUp(self):
-        self.universe = MDAnalysis.Universe(TPR,XTC)
+class TestSelectionsTPR(object):
+    @staticmethod
+    @pytest.fixture
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/core/test_atomselections.py]: iirc, tpr is a little slow, can we expand scope of this for perf?

File: package/MDAnalysis/core/groups.py (hunk 1 above)
Comment (jbarnoud) [inline package/MDAnalysis/core/groups.py]: I did hesitate. On one hand it is indeed an attribute missing that cause the problem; on the other hand, on the user point of view, it is the argument input that is the problem. I'll fix it.

**Comment by richardjgowers** [comment]
@jbarnoud ok sure, but virtual sites not getting properly treated is a failure of the bond-walking algorithm, rather than the difference between what you could call a fragment and what you could call a molecule.  I'm thinking it'd be nicer to condense these two concepts so we have less nomenclature floating around.

File: testsuite/MDAnalysisTests/core/test_atomselections.py (hunk 2 above)
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/core/test_atomselections.py]: I'll try.

**Comment by jbarnoud** [comment]
@richardjgowers What I worry about is having `fragments` behaving differently depending on the format. I see only one use case where it would actually matter (the bond pruning described above), and that use case is very theoretical. Yet, I would expect a given behavior from `fragments` (actual connected components) when having a different one because of my input format (annotated molecules).

I am biased, though, as I know the semantic difference between a fragment and a molecule. So I may underestimate the confusion of having the two concepts...

**Comment by richardjgowers** [comment]
Ok thinking about this, I'd like to merge the concepts of molecules and fragments.  Most formats will guess molecules (based on bonding) this format will define molecules.

**Comment by richardjgowers** [comment]
But I'm happy to let people disagree.  This would mirror the behaviour for masses & bonds, where it is sometimes guessed, sometimes read from topology.  As long as proper warnings are in place it should be fine.

What I'm trying to avoid is having many methods accepting either `fragments` or `molecules` as keywords (eg things like `split()`) when really they refer to the same concept.

**Comment by jbarnoud** [comment]
I had some thought about it and it makes sense. I am only concerned 
about the specific use case of removing a bond: then you get two 
fragments, but the molecule information you read from the topology does 
not translate it.

**Comment by orbeckst** [comment]
Just having "molecules" would be better and less confusing than having two closely related almost always identical concepts. A "molecule" is ultimately going to be whatever we choose to represent a real chemical molecule by although it should be close to _"a set of atoms that is connected by covalent bonds is a molecule"_ (replaced atoms by particles for coarse grained force fields).

There can be different ways to arrive at this definition
- labeling with a molnum (take the ground truth from the software that build it) – that's probably the best approach because then we don't have to do any chemistry or guess the chemistry
- guessing by following covalent bonds and using the (pseudo)chemical definition 

If you make it easy and show how one could, for instance, manually add v-sites to a bond-guessed molecule or remove v-sites (do they have any special labels in the TPR and do we expose these labels?) the everyone could create the kind of "molecules/fragments" that @jbarnoud mentions in his use case.

My $0.02.

**Comment by orbeckst** [comment]
> I am only concerned about the specific use case of removing a bond: then you get two fragments, but the molecule information you read from the topology does not translate it.

Maybe this is not our problem if a user decides to start topology hacking? I mean, perhaps it need not be supported at the core level but instead handled like `Merge()`: have helper functions that "do the right thing", like a tool box of topology modifying functions. A `cut_bond()` would then explicitly check connectivity and return a list of 1 or 2 molecules.

Or is the use case general enough that we can robustly implement it?

**Comment by richardjgowers** [comment]
Ok so to recap this PR so far... @jbarnoud added the ability to read "molecules" from a TPR (awesome!)

It then got a little sticky as I didn't want to introduce "molecules" in addition to "fragments" and wanted to merge them to have fewer concepts flying around.  The downside of this being that sometimes a "molecule" wouldn't always be continuously bonded in corner cases (where a VSite is holding the molecule together).

So options:
 - 1 add molecules in addition to fragments.
    - Pros: fragments are *always* continuously connected
    - Cons: another attribute for functions to deal with, frags/molecules are *very* similar 
 - 2 merge molecules and fragments
    - Pros: fewer keywords - keeps things conceptually tight.
    - Cons: breaks assumption that fragments have continuous bonds

I think I'm for option 2, I think our current method of doing fragments is just an approximation and if a topology file (here TPR) provides fragment information we can and should use that.  I think GSD also does fragment information now I think about it...

@MDAnalysis/coredevs opinions?

**Comment by orbeckst** [comment]
I am also in favor of 2 (see my comments above). But unifying fragments and molecules means breaking the old fragments behavior and either deprecating or switching cold-turkey.

@richardjgowers how would you want to do the transition to molecules only? Deprecated `fragments` or making `fragments` and alias of `molecules`?

**Comment by kain88-de** [comment]
Where can I find the fragment code and how did we document this so far? I want to understand how we defined it because I was in aware this even existed

**Comment by kain88-de** [comment]
I've looked into the code and docs. We don't really define what a fragment is. With this move we could define it and move to combine it with the molecule keyword (which I find better then fragments). I'm therefore also for option 2.

For option 2 the way forward would be to move `fragments`/`molecule` from an [adhoc addition](https://github.com/MDAnalysis/mdanalysis/blob/699697a45e0177a4cb12df67a5fd692b8ba9cbbe/package/MDAnalysis/core/universe.py#L844) of a universe to a real topology attribute? This would entail:

1. Remove `fragments` from universe
1. make `fragments` an alias of `molecule` with deprecation warning
1. add `molecule` to topologies with bond information (PDB, PSF, ???)

would the guess method be done at parsing of the topology or lazily upon requesting the property?

**Comment by jbarnoud** [comment]
Sorry, I've been rather quiet these days.

I know `residues` is used quite a lot, at least in my lab. So I would avoid breaking its current behaviour too much.

What I am planning on doing is to have `fragments` and `molecules` be aliases of each other. If `fragments` find the `molnum` attribute, it uses it to build  `fragment` and `fragments` for the groups; else it keeps the current behaviour.

I would add a 4th option to @kain88-de's list:

4. just keep the name `fragments` and do not create a `molecules` attribute.

**Comment by richardjgowers** [comment]
Yeah Fragments is currently inside the Bond TopAttr atm which isn't ideal if it can have one of two constructions. It might make most sense to attach Fragments to Universe always and give it similar status to Residues and Segments; always there but sometimes ignored by topologies.

Giving priority to molnum sounds reasonable. 

We can't only add it to some formats as you can guess Bonds (eg from xyz) and then you would have enough info. 

Aliasing the two sounds right.

**Comment by jbarnoud** [comment]
I rebased the PR and made the tests pass. The molnums feature itself is done. Maybe merging fragments and molecules can be a separate PR.

**Comment by kain88-de** [comment]
I want to add another benefit of this PR. Trying to calculate fragments for a 1.3 million atoms systems (TPR) is consuming several hundred megabytes of RAM, this should be much better with molid.

**Comment by richardjgowers** [comment]
We can refine the difference between Fragments and Molecules next release

**Pull request merged**